### PR TITLE
Alias for arguments

### PR DESCRIPTION
**Pull request by crissi**
especially useful for mutation where you would like the inputs to match the database columns

### CODE REVIEW

**Comment by mfn** [comment]
I'm not against this feature.

But I don't like that it's _forced_ upon everyone.

We support the concept of `'alias'` on fields but this is optional: only if you make use of `$getSelectFields()` you opt-in to that processing and the overhead of it.

Not so with this approach as it's hardcoded in the resolver (and I would find it a bit extreme to override the resolver just to be able to disable this).

It feels this is in a similar space as the idea of https://github.com/rebing/graphql-laravel/issues/505 , namely to activate some optional kind of processing in various stages here. I think this belongs in a bigger picture (which I don't have at this moment).

**Comment by crissi** [comment]
> I'm not against this feature.
> 
> But I don't like that it's _forced_ upon everyone.
> 
> We support the concept of `'alias'` on fields but this is optional: only if you make use of `$getSelectFields()` you opt-in to that processing and the overhead of it.
> 
> Not so with this approach as it's hardcoded in the resolver (and I would find it a bit extreme to override the resolver just to be able to disable this).
> 
> It feels this is in a similar space as the idea of #505 , namely to activate some optional kind of processing in various stages here. I think this belongs in a bigger picture (which I don't have at this moment).

I see you point, but I been using this for quite a time right now and I think it such a common usecase, especially for mutations.

But it could also be implemented as as middleware of some sort.

**Comment by mfn** [comment]
> But it could also be implemented as as middleware of some sort.

I was thinking in that direction but when I look at the PR, if I understand it correctly, the changes have to be done _after_ validation but _before_ invoking the resolver (actually not even sure why/how it works correctly with the $getSelectFields closure).

So this would require a *very* specific _kind of_ middleware, like "RightInTheMiddleOfGetResolverMiddleware" 😜 or something.

This middleware thing would have to receive and return all arguments but should probably just receive everything, maybe even `$this->type()`.

> I been using this for quite a time right now and I think it such a common usecase, especially for mutations.

How did you go about it? Did you override `\Rebing\GraphQL\Support\Field::getResolver`?

**Comment by crissi** [comment]
> > But it could also be implemented as as middleware of some sort.
> 
> I was thinking in that direction but when I look at the PR, if I understand it correctly, the changes have to be done _after_ validation but _before_ invoking the resolver (actually not even sure why/how it works correctly with the $getSelectFields closure).
> 
That one still receives the original input
> So this would require a _very_ specific _kind of_ middleware, like "RightInTheMiddleOfGetResolverMiddleware" 😜 or something.

:grin:

> 
> This middleware thing would have to receive and return all arguments but should probably just receive everything, maybe even `$this->type()`.
> 
> > I been using this for quite a time right now and I think it such a common usecase, especially for mutations.
> 
> How did you go about it? Did you override `\Rebing\GraphQL\Support\Field::getResolver`?

yes

**Comment by mfn** [comment]
> > How did you go about it? Did you override `\Rebing\GraphQL\Support\Field::getResolver`?
> 
> yes

Ok, what I thought / feared 😄  But it highlights the shortcomings, thanks.

> That one still receives the original input

But I then think this isn't quite appropriate, per you initial description:
> where you would like the inputs to match the database columns

This would mean the actually `SelectFields` would **not** benefit from that.


----

Maybe it needs some more hooks into the system or, as I stated already, looking into GraphQL directives. But I don't want to incur overhead for a feature, as useful as it sounds, which may not be used at all by consumers of the library (aka opt-in).

My TL;DR is that I don't think modifying the resolver this way is the right approach and as such I don't want to accept the PR (at least in its current stage).

**Comment by crissi** [comment]
> > > How did you go about it? Did you override `\Rebing\GraphQL\Support\Field::getResolver`?
> > 
> > 
> > yes
> 
> Ok, what I thought / feared 😄 But it highlights the shortcomings, thanks.
> 
> > That one still receives the original input
> 
> But I then think this isn't quite appropriate, per you initial description:
> 
> > where you would like the inputs to match the database columns
> 
> This would mean the actually `SelectFields` would **not** benefit from that.
> 

I guess this could be added here to, I just never actual thought about that since I do not use the https://github.com/rebing/graphql-laravel#type-relationship-query very often

> Maybe it needs some more hooks into the system or, as I stated already, looking into GraphQL directives. But I don't want to incur overhead for a feature, as useful as it sounds, which may not be used at all by consumers of the library (aka opt-in).
> 
> My TL;DR is that I don't think modifying the resolver this way is the right approach and as such I don't want to accept the PR (at least in its current stage).

fair

**Comment by crissi** [comment]
Should I close this or would you accept this with the modifications to the selectFields?

of course an option that enables this could also be added?

**Comment by mfn** [comment]
Thanks for asking, I didn't close it as I didn't want to signal an outright rejection, just the implementation.

Whatever fits you, re-iterating this PR or making a new one, I don't mind either.

**Comment by crissi** [comment]
but you want one that does not modify the getResolver directly?

**Comment by mfn** [comment]
Not per se, I try to formulate what I think we should be aware of:
- keep the API simple
- don't force it on users if not needed

I already think it's not really elegant that `resolve` receives a `$getSelectFields` argument and which there was another way, but here we already tried to remove hardcoded "opt-in". Before that change everyone was always getting a `SelectFields` class even if you didn't use it.

OTOH when I think about it, a feature like supporting aliases for input argument might not even something that should be easily turned off / on. It can be useful regardless of the database context.

But when I look at the current implemention `AliasArguments`, it already has to do quite some work and that is even only half of the story, because parts are in the external dependency.

**Comment by georgeboot** [comment]
I would also love to see a feature like this.
My database columns are all `snake_case`, whereas the fields in my types are all `camelCase`.

I now find my self constantly mapping them by hand, whereas a `new Model($args['input'])` in combination with `$fillable` would be much easier.

Regarding the added overhead: have you done any research on how big the overhead is?
In my opinion, If te API remains simple, I don't think we should hold this off because added code.

**Comment by mfn** [comment]
Hmm. Is there an easier way to achieve this without the this PRs `AliasArguments` and it's dependency? Just a gut feeling but it seems a lot of logic necessary for this.

(note: please don't be offended; I haven't attempted an implementation myself, just thought I'd ask…)

**Comment by georgeboot** [comment]
There might be.

What is your main concern? Added dependency, performance (can’t easily opt-out or this code) or complexity?

**Comment by mfn** [comment]
Haha, all :-)

I think I'm overruled with my performance concerns in that a) there's more need for hit and b) so far the concern is only on paper, no one (including myself) has proved it.

I would definitely prefer a dependency-less solution. Not because I'm against reusing code, but it seems to me this package was created with the almost sole purpose here. I.e. it would make more sense to me to have this in this code base, for easier review and future optimizations.

**Comment by crissi** [comment]
I will include the dependency and refactor a bit too see if we can get closer to an acceptable solution.

**Comment by mfn** [comment]
>  too see if we can get closer to an acceptable solution.

Thank you very much!

I know that some of my feedback here is very … non-argumentative on a technical level; and on top not quite so specific. Would appreciate a take from a (slightly) different angle here; thx.

**Comment by mfn** [comment]
@crissi is this ready for review?

**Comment by crissi** [comment]
> @crissi is this ready for review?

yes

File: src/Support/Field.php [hunk 1]
```diff
@@ -204,6 +205,8 @@ protected function getResolver(): ?Closure
                 };
             }
 
+            $arguments[1] = (new AliasArguments($this->args(), $arguments[1]))->get();
```
Comment (mfn) [inline src/Support/Field.php]: How about this:
- we add a new `protected function aliasArgs(array $args): array`
  ```php
  {
    return (new AliasArguments($this->args(), $arguments[1]))->get();
  }
  ```
- we add a new `protected function getArgs(array $args): array`
  ```php
  {
    return $this->aliasArgs($args);
  }
  ```
- we call it here like this: `$arguments[1] = $this->getArgs($arguments[1]);`

I see the benefit that:
- people can decide what to do with the args on their own
- by providing `aliasArgs` we make it easier to re-use if it people want to add additional steps within their `getArgs()`

File: tests/Unit/AliasAguments/Stubs/UpdateExampleMutation.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
+<?php
```
Comment (mfn) [inline tests/Unit/AliasAguments/Stubs/UpdateExampleMutation.php]: declare strict_types missing

File: tests/Unit/AliasAguments/Stubs/ExampleValidationInputObject.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<?php
```
Comment (mfn) [inline tests/Unit/AliasAguments/Stubs/ExampleValidationInputObject.php]: declare strict_types missing

File: tests/Unit/AliasAguments/Stubs/ExampleType.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<?php
```
Comment (mfn) [inline tests/Unit/AliasAguments/Stubs/ExampleType.php]: declare strict_types missing

File: tests/Unit/AliasAguments/Stubs/ExampleNestedValidationInputObject.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
+<?php
```
Comment (mfn) [inline tests/Unit/AliasAguments/Stubs/ExampleNestedValidationInputObject.php]: declare strict_types missing

File: tests/Unit/AliasAguments/AliasArgumentsTest.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
+<?php
```
Comment (mfn) [inline tests/Unit/AliasAguments/AliasArgumentsTest.php]: declare strict_types missing

**Comment by crissi** [comment]
@mfn can you look again, I see styleci is still active?!

**Comment by crissi** [comment]
@mfn can you take a look again?

**Comment by crissi** [comment]
styleci still running!?

**Comment by mfn** [comment]
@crissi 
Ok, I guess I now see were the confusion also from the other PR came.

Please remove the changes to please styleci => we can ignore the checks by it until it's removed

sorry for the confusion.

**Comment by mfn** [comment]
Please just remove/revert the commits you made to please StyleCI, that should be the easiest way AFAICS.

**Comment by crissi** [comment]
@mfn should be ready for review now

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 --------------
 ### Added
 - Allow passing through an instance of a `Field` [\#521 / georgeboot](https://github.com/rebing/graphql-laravel/pull/521/files)
+- add the ability to alias query and mutations arguments as well as input objects [\#517 / crissi](https://github.com/rebing/graphql-laravel/pull/517/files)
```
Comment (mfn) [inline CHANGELOG.md]:
```suggestion
- Add the ability to alias query and mutations arguments as well as input objects [\#517 / crissi](https://github.com/rebing/graphql-laravel/pull/517/files)
```

File: Readme.md [hunk 8]
```diff
@@ -101,7 +101,7 @@ To work this around:
 ## Usage
 
 - [Laravel GraphQL](#laravel-graphql)
-    - [Note: these are the docs for 3.*, please see the `v1` branch for the 1.* docs](#note-these-are-the-docs-for-3-please-see-the-v1-branch-for-the-1-docs)
+    - [Note: these are the docs for the current release, please see the `v1` branch for the 1.* docs](#note-these-are-the-docs-for-the-current-release-please-see-the-v1-branch-for-the-1-docs)
```
Comment (mfn) [inline Readme.md]: How is this diff even possible? The master branch literally says this: https://github.com/rebing/graphql-laravel/commit/30e69d32ada916ae5f7caeca4783c0380f167aa7

Comment (crissi) [reply Readme.md]: probably my markdown editor that changed that, will revert

**Comment by mfn** [comment]
> I'll take care of the broken tests, this is a travis config issue.

https://github.com/rebing/graphql-laravel/pull/538

File: Readme.md (hunk 8 above)
Comment (mfn) [reply Readme.md]: No, I mean the version in master already is at the change so technically this shouldn't be possible.

No need to change it, it's the _correct_ text anyway!

**Comment by mfn** [comment]
Awesome, thank you!

**Pull request merged**